lib/ziti-tunnel-cbs/ziti_hosting.c

```diff
@@ -205,14 +213,45 @@ static ssize_t on_hosted_client_data(ziti_connection clt, uint8_t *data, ssize_t
     return len;
 }
 
-#define ZITI_MTU (15 * 1024)
 static void alloc_buffer(uv_handle_t *handle, size_t suggested_size, uv_buf_t *buf) {
-    *buf = uv_buf_init((char*) malloc(ZITI_MTU), ZITI_MTU);
+    struct hosted_io_ctx_s *io_ctx = handle->data;
+
+    if (io_ctx->in_wreqs < MAX_OUTSTANDING_WRITES) {
+        *buf = uv_buf_init((char*) malloc(ZITI_MTU), ZITI_MTU);
+    } else {
+        ZITI_LOG(VERBOSE, "max ziti writes[%d] reached for %s", io_ctx->in_wreqs, io_ctx->server_dial_str);
+        // provoke UV_ENOBUFS
+        buf->base = NULL;
+        buf->len = 0;
+    }
 }
 
 /** called by ziti SDK when data transfer initiated by ziti_write completes */
 static void on_hosted_ziti_write(ziti_connection ziti_conn, ssize_t len, void *ctx) {
-    free(ctx);
+    struct hosted_io_ctx_s *io_ctx = ziti_conn_data(ziti_conn);
+
+    if (ctx) free(ctx);
+
+    io_ctx->in_wreqs--;
+    if (io_ctx->in_wreqs < 0) {
+        ZITI_LOG(ERROR, "WTF: accounting error");
+        io_ctx->in_wreqs = 0;
+    }
+
+    if (len < 0) {
+        ZITI_LOG(WARN, "ziti write error: %zd(%s), stop reading peer", len, ziti_errorstr((int)len));
+
+        switch (io_ctx->server_proto_id) {
+            case IPPROTO_TCP:
+                uv_read_stop((uv_stream_t *) &io_ctx->server.tcp);
+                break;
+            case IPPROTO_UDP:
+                uv_udp_recv_stop(&io_ctx->server.udp);
+                break;
```

Comment:
does it makes sense to have a 'guard' case should we ever do something that is not tcp/udp?